sbi/inference/posteriors/vi_posterior.py

```diff
@@ -1,69 +1,481 @@
 # This file is part of sbi, a toolkit for simulation-based inference. sbi is licensed
 # under the Affero General Public License v3, see <https://www.gnu.org/licenses/>.
-from functools import partial
-from math import ceil
-from typing import Any, Callable, Dict, List, Optional, Union
+
+from copy import deepcopy
+from typing import Callable, List, Optional, Union, Dict
 from warnings import warn
 
 import numpy as np
 import torch
-import torch.distributions.transforms as torch_tf
-from pyro.infer.mcmc import HMC, NUTS
-from pyro.infer.mcmc.api import MCMC
-from torch import Tensor, nn
+from torch.distributions import Distribution
+
+from tqdm import tqdm
+
 
-from sbi import utils as utils
 from sbi.inference.posteriors.base_posterior import NeuralPosterior
-from sbi.samplers.rejection.rejection import rejection_sample
+from sbi.utils import mcmc_transform
 from sbi.types import Shape, TorchTransform
-from sbi.utils import del_entries
-from sbi.utils.torchutils import atleast_2d, ensure_theta_batched
+from sbi.utils.torchutils import (
+    atleast_2d_float32_tensor,
+    Array,
+    Tensor,
+    ensure_theta_batched,
+)
+
+from sbi.samplers.vi import (
+    adapt_variational_distribution,
+    check_variational_distribution,
+    make_object_deepcopy_compatible,
+    move_all_tensor_to_device,
+    get_flow_builder,
+    get_VI_method,
+    get_sampling_method,
+    get_quality_metric,
+    get_default_flows,
+    get_default_VI_method,
+    get_default_sampling_methods,
+    get_sampling_method_parameters_doc,
+    docstring_parameter,
+)
 
 
 class VIPosterior(NeuralPosterior):
-    r"""Posterior $p(\theta|x)$ with `log_prob()` and `sample()` methods, obtained with
-    SNLE.<br/><br/>
-    SNLE trains a neural network to approximate the likelihood $p(x|\theta)$. The
-    `SNLE_Posterior` class wraps the trained network such that one can directly evaluate
-    the unnormalized posterior log probability $p(\theta|x) \propto p(x|\theta) \cdot
-    p(\theta)$ and draw samples from the posterior with MCMC.<br/><br/>
-    The neural network itself can be accessed via the `.net` attribute.
+    r"""Provides VI (Variational Inference) to sample from the posterior.<br/><br/>
+
+    SNLE or SNRE train neural networks to approximate the likelihood(-ratios).
+    `VIPosterior` allows to train a tractable variational posterior q(theta) which
+    approximates the true posterior p(theta|x_o). After this second training stage, we
+    can produce approximate posterior samples, by just sampling from q with no
+    additional cost.
     """
 
+    @docstring_parameter(get_default_flows(), get_default_VI_method())
     def __init__(
         self,
         potential_fn: Callable,
-        prior,
+        q: Union[str, Distribution, NeuralPosterior] = "maf",
```

Comment:
Yes, I just used NeuralPosterior because PyLance complains otherwise (as VIPosterior is not defined at this moment). But I can change it, makes much more sense.

Reply:
A

Reply:
Ah it actually also throws an "NameError", so cannt realy be done. I may can add a check within the set_q

Reply:
In the type hint, you can just put quotation marks around the `VIPosterior`, i.e. `Union[str, Distribution, "VIPosterior"]

Reply:
Ah cool thanks!